docs/graalvm-compatibility-in-ballerina-libraries.md

```diff
@@ -0,0 +1,185 @@
+
+2. A text editor
+   >**Tip:** Preferably, <a href="https://code.visualstudio.com/" target="_blank">Visual Studio Code</a> with the  <a href="https://wso2.com/ballerina/vscode/docs/" target="_blank">Ballerina extension</a> installed.
+
+3. GraalVM installed and configured appropriately
+   
+#### Configure GraalVM locally
+1. Install GraalVM using [SDKMAN!](https://sdkman.io/). 
+   >**Tip:** For additional download options, see [Get Started with GraalVM](https://www.graalvm.org/jdk17/docs/getting-started/).
+
+      ```
+      $ sdk install java 17.0.7-graalce
+      ```
+      > **Note:** If you have installed Ballerina Swan Lake Update 7(2201.7.x) or lower, you have to install GraalVM JDK 11. For download options, see [Get Started with GraalVM](https://www.graalvm.org/22.3/docs/getting-started/macos/).
+      
+2. Set the `GRAALVM_HOME` environment variable to the GraalVM installation directory. If you have installed using SDKMAN! you can set it to `JAVA_HOME`.
+
+> **Note:** 
+> - On Windows, the native image requires Visual Studio Code and Microsoft Visual C++ (MSVC). For more details, see [Prerequisites for Native Image on Windows](https://www.graalvm.org/latest/docs/getting-started/windows/#prerequisites-for-native-image-on-windows).
+> - The GraalVM native-image tool support for Apple M1 (darwin-aarch64) is still experimental. For more updates, see [Support for Apple M1](https://github.com/oracle/graal/issues/2666).
+
+#### Build with GraalVM
+1. Run the following command to build a native executable for the sample application
+   ```
+   $ bal build --graalvm
+   ```
+   
+2. Run the executable generated in the `target/bin` directory
+   ```
+   $ ./target/bin/sample
+   ```
+
+#### Test with GraalVM
+1. Add Ballerina test cases for the sample application
+  
+2. Run the following command to run the tests in a native executable
+   ```
+   $ bal test --graalvm
+   ```
+
+#### Handle errors
+There may be errors when building the native image due to [class initialization](https://www.graalvm.org/jdk17/reference-manual/native-image/optimizations-and-performance/ClassInitialization/). Fix these errors using the error logs and tracing the class initialization. For more information, see [Updates on Class Initialization in GraalVM Native Image Generation](https://medium.com/graalvm/updates-on-class-initialization-in-graalvm-native-image-generation-c61faca461f7).
+
+Even though the build passes, running the executable may end up in unexpected errors. This could happen if we have not added all the necessary configurations related to the Java dynamic features. The necessary configurations needed for this particular sample application can be automatically found by engaging the [tracing agent](https://www.graalvm.org/jdk17/reference-manual/native-image/metadata/AutomaticMetadataCollection/) when running the jar file.
```

Comment:
```suggestion
Even though the build passes, running the executable may end up in unexpected errors. This could happen if you have not added all the necessary configurations related to the Java dynamic features. The necessary configurations needed for this particular sample application can be automatically found by engaging the [tracing agent](https://www.graalvm.org/jdk17/reference-manual/native-image/metadata/AutomaticMetadataCollection/) when running the jar file.
```